modeshape-jcr/src/main/java/org/modeshape/jcr/txn/Transactions.java

```diff
@@ -384,11 +384,12 @@ public void commit()
 
     protected class NestableThreadLocalTransaction extends TraceableSimpleTransaction {
         private AtomicInteger nestedLevel = new AtomicInteger(0);
-        private ThreadLocal<? extends Transaction> txReference;
+        private ThreadLocal<NestableThreadLocalTransaction> txHolder;
```

Comment:
Hmm.... If a `NestableThreadLocalTransaction` is created in a certain thread, then by using `ThreadLocal` here the code below will only find a non-null reference when accessed from the same thread. If the transaction is resumed in a different thread, upon cleanup (when the code tries to clean up this `NestableThreadLocalTransaction`) the contents of the `txHolder` will not be visible to the current thread. I've not got my head wrapped around this, but it seems like we're not properly removing the reference in `txHolder` so that when a new transaction is started on the same thread that started our first transaction, it will find an existing `NestableThreadLocalTransaction` reference.

Having `NestableThreadLocalTransaction` hold onto the thread local seemed like a good idea when it was assumed that a local transaction would only be active on a single thread. But now that we know that's not a valid assumption, this class should no longer hold onto the thread local.

Something just seems not quite right.